thunder/distributed/__init__.py

```diff
@@ -349,6 +350,162 @@ def f(tensor: TensorProxy) -> str:
+        module_copy = copy.copy(submodule)
+
+        # Materialize meta-parameters on-device if necessary.
+        # This is done before sharding in case the materialization logic depends on the tensor shape.
+        # The tradeoff is that all of a module's direct parameters need to fit in device.
+        # Each module only initializes its own parameters and not those of its children (recurse=False)
+        if any(t.is_meta for t in chain(module_copy.parameters(recurse=False), module_copy.buffers(recurse=False))):
+            # TODO: we could also support calling a "param_init_fn" argument like PyTorch
+            # thunder.distributed._materialize(submodule, device)
+            # Materialize a module's direct children parameters by calling ``module.reset_parameters()``
+            module_copy.to_empty(device=device, recurse=False)
+            if not hasattr(module_copy, "reset_parameters"):
+                raise TypeError(
+                    f"Materialization requires that the `{type(module_copy).__name__}.reset_parameters` method is implemented."
+                    " This method is used to initialize any children parameters or buffers in this module."
+                )
+            module_copy.reset_parameters()
+            for n, p in module_copy.named_parameters(recurse=False, prefix=module_name):
+                thunder_model._overrides[n] = p
+                device_adjustments[n] = device
+            for n, b in module_copy.named_buffers(recurse=False, prefix=module_name):
+                thunder_model._overrides[n] = b
+                device_adjustments[n] = device
+        else:
+            # Move leftover params and buffers to device. This is at least required to broadcast.
+            # Cannot `submodule.to(device)` because we don't want it to recurse
+            for n, p in module_copy.named_parameters(recurse=False, prefix=module_name):
+                if p.device != device:
+                    thunder_model._overrides[n] = torch.nn.Parameter(p.to(device=device), requires_grad=p.requires_grad)
+                    device_adjustments[n] = device
+            for n, b in module_copy.named_buffers(recurse=False, prefix=module_name):
+                if p.device != device:
+                    thunder_model._overrides[n] = b.to(device=device)
+                    device_adjustments[n] = device
+
+        for pn, p in submodule.named_parameters(recurse=False, prefix=module_name):
+            if pn not in thunder_model._overrides:
+                thunder_model._overrides[pn] = copy.copy(p)
+            # we collect shapes and devices because we do not know if other transforms also change it...
+            old_shape = thunder_model._overrides[pn].shape
+            thunder.distributed._shard_param(thunder_model._overrides[pn], global_rank, world_size, pn)
+            new_shape = thunder_model._overrides[pn].shape
+            sharded_params[pn] = (old_shape, new_shape, thunder_model._overrides[pn].device)
```

Comment:
Where is this information used?

Reply:
This is used in the `prologue_and_compute_transform` callback. The thing is that we need to transform the `ThunderModule` when the fsdp(...) is applied but the `prologue_and_compute_transform` needs to run when the ThunderModule is actually called (and we have a cache miss).